File: java/33.java
class Solution {
    public int search(int[] nums, int target) {
        return searchRecursively(nums, target, 0, nums.length - 1);
    }

    public int searchRecursively(int[] nums, int target, int start, int end) {
        if (start > end) {
            return -1;
        }
        int mid = (start + end) / 2;
        if (nums[mid] == target) {
            return mid;
        }
        int result = -1;
        if (((nums[start] < nums[mid]) && (nums[start] <= target && target <= nums[mid])) || (nums[start] > nums[mid])) {
            int leftResult = searchRecursively(nums, target, start, mid - 1);
            if (leftResult != -1) {
                result = leftResult;
            }
        }
        if (((nums[mid] < nums[end]) && (nums[mid] <= target && target <= nums[end])) || (nums[mid] > nums[end])) {
            int rightResult = searchRecursively(nums, target, mid + 1, end);
            if (rightResult != -1) {
                result = rightResult;
            }
        }
        return result;
    }
}

File: python/33.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.searchRecursively(nums, target, 0, len(nums) - 1)
    
    def searchRecursively(self, nums: List[int], target: int, left: int, right: int) -> int:
        if left == right:
            if target == nums[left]:
                return left
            else:
                return -1
        if left > right:
            return -1
        mid = int((left + right) / 2)
        if nums[mid] == target:
            return mid
        result = -1
        if nums[left] > nums[mid]:
            curr_result = self.searchRecursively(nums, target, left, mid - 1)
            if curr_result != -1:
                result = curr_result
        else:
            if nums[left] <= target and target < nums[mid]:
                curr_result = self.searchRecursively(nums, target, left, mid - 1)
                if curr_result != -1:
                    result = curr_result

        if nums[mid] > nums[right]:
            curr_result = self.searchRecursively(nums, target, mid + 1, right)
            if curr_result != -1:
                result = curr_result
        else:
            if nums[mid] < target and target <= nums[right]:
                curr_result = self.searchRecursively(nums, target, mid + 1, right)
                if curr_result != -1:
                    result = curr_result
        return result
